include/cuco/detail/bloom_filter_kernels.cuh

```diff
@@ -0,0 +1,108 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace cuco {
+namespace detail {
+namespace cg = cooperative_groups;
+
+/**
+ * @brief Initializes each slot in the flat `slot` storage.
+ *
+ * @tparam block_size The size of the thread block
+ * @tparam atomic_slot_type Type of the slot's atomic container
+ * @param slots Pointer to flat `slot` storage
+ * @param num_slots Size of the storage pointed to by `slots`
+ */
+template <std::size_t block_size, typename atomic_slot_type>
+__global__ void initialize(atomic_slot_type* const slots, std::size_t num_slots)
+{
+  for (std::size_t tid = block_size * blockIdx.x + threadIdx.x; tid < num_slots;
+       tid += gridDim.x * block_size) {
+    new (&slots[tid]) atomic_slot_type{0};
+  }
+}
+
+/**
+ * @brief Inserts all keys in the range `[first, last)`.
+ *
+ * @tparam block_size The size of the thread block
+ * @tparam InputIt Device accessible input iterator whose `value_type` is
+ * convertible to the filter's `key_type`
+ * @tparam View Type of device view allowing access of filter storage
+ * @tparam Hash Unary callable type
+ * @param first Beginning of the sequence of keys
+ * @param last End of the sequence of keys
+ * @param view Mutable device view used to access the filter's slot storage
+ * @param hash The unary function to apply to hash each key
+ */
+template <std::size_t block_size, typename InputIt, typename View, typename Hash>
+__global__ void insert(InputIt first, InputIt last, View view, Hash hash)
+{
+  std::size_t tid = block_size * blockIdx.x + threadIdx.x;
+  auto it         = first + tid;
+
+  while (it < last) {
+    typename View::key_type const key{*it};
+    view.insert(key, hash);
+    it += gridDim.x * block_size;
+  }
+}
+
+/**
+ * @brief Indicates whether the keys in the range `[first, last)` are contained
+ * in the filter.
+ *
+ * Writes a `bool` to `(output + i)` indicating if the key `*(first + i)` exists
+ * in the filter.
+ *
+ * @tparam block_size The size of the thread block
+ * @tparam InputIt Device accessible input iterator whose `value_type` is
+ * convertible to the filter's `key_type`
+ * @tparam OutputIt Device accessible output iterator whose `value_type` is
+ * convertible to `bool`.
+ * @tparam View Type of device view allowing access of filter storage
+ * @tparam Hash Unary callable type
+ * @param first Beginning of the sequence of keys
+ * @param last End of the sequence of keys
+ * @param output_begin Beginning of the sequence of booleans for the presence of each key
+ * @param view Mutable device view used to access the filter's slot storage
+ * @param hash The unary function to apply to hash each key
+ */
+template <std::size_t block_size, typename InputIt, typename OutputIt, typename View, typename Hash>
+__global__ void contains(InputIt first, InputIt last, OutputIt output_begin, View view, Hash hash)
```

Comment:
Specifying launch bounds sometimes affects performance:
```
__global__ void __launch_bounds__(block_size)
```

Reply:
I just tested this but sadly there is no performance improvement. Maybe the compiler is already smart enough to deduce the bound from the `block_size` tparam. Or we are simply not limited by occupancy. I need to run a profile to find out.

Reply:
It's often just "good practice" to specify the `__launch_bounds__` when the CTA size is known statically. Even if it doesn't impact perf on the benchmarks we have on the specific hardware you tested, there may still be situations (present or future) where it does help.